server/debug/src/symbol.rs

```diff
@@ -0,0 +1,254 @@
+        unsafe {
+            let mut displacement = 0;
+            let mut symbol = (winapi::SYMBOL_INFOW {
+                SizeOfStruct: mem::size_of::<winapi::SYMBOL_INFOW>() as winapi::ULONG,
+                MaxNameLen: winapi::MAX_SYM_NAME as winapi::ULONG,
+                ..mem::zeroed()
+            }, [0 as winapi::WCHAR; (winapi::MAX_SYM_NAME - 1) as usize]);
+
+            if dbghelp::SymFromAddrW(
+                self.0, address as winapi::DWORD64, &mut displacement, &mut symbol.0
+            ) == winapi::FALSE {
+                return Err(io::Error::last_os_error());
+            }
+
+            let name = slice::from_raw_parts(symbol.0.Name.as_ptr(), symbol.0.NameLen as usize);
+            let name = OsString::from_wide(name);
+
+            Ok((Symbol { name, address: symbol.0.Address as usize }, displacement as usize))
+        }
+    }
+
+    /// Retrieve the filename, line number, and byte offset of an instruction address
+    pub fn line_from_address(&mut self, address: usize) -> io::Result<(OsString, u32, usize)> {
+        unsafe {
+            let mut displacement = 0;
+            let mut line = winapi::IMAGEHLP_LINEW64 {
+                SizeOfStruct: mem::size_of::<winapi::IMAGEHLP_LINEW64>() as winapi::DWORD,
+                ..mem::zeroed()
+            };
+
+            if dbghelp::SymGetLineFromAddrW64(
+                self.0, address as winapi::DWORD64, &mut displacement, &mut line
+            ) == winapi::FALSE {
+                return Err(io::Error::last_os_error());
+            }
+
+            let mut len = 0;
+            while *line.FileName.offset(len as isize) != 0 { len += 1; }
+
+            let name = slice::from_raw_parts(line.FileName, len);
+            let name = OsString::from_wide(name);
+
+            Ok((name, line.LineNumber, displacement as usize))
+        }
+    }
+
+    /// Iterate through the frames of a thread's stack
+    ///
+    /// The thread should be part of an attached child process which is currently paused to handle
+    /// a debug event.
+    pub fn walk_stack(&mut self, thread: winapi::HANDLE) -> io::Result<StackFrames> {
+        unsafe {
+            let mut context = winapi::CONTEXT {
+                ContextFlags: winapi::CONTEXT_FULL,
+                ..mem::zeroed()
+            };
+            if kernel32::GetThreadContext(thread, &mut context) == winapi::FALSE {
+                return Err(io::Error::last_os_error());
+            }
+
+            fn flat(address: winapi::DWORD64) -> winapi::ADDRESS64 {
+                winapi::ADDRESS64 { Offset: address, Mode: winapi::AddrModeFlat, Segment: 0 }
+            }
+            let stack = winapi::STACKFRAME64 {
+                AddrPC: flat(context.Rip),
+                AddrFrame: flat(context.Rbp),
+                AddrStack: flat(context.Rsp),
+                ..mem::zeroed()
+            };
+
+            Ok(StackFrames { process: self.0, thread, context, stack })
+        }
+    }
```

Comment:
Is the reason we don't see an iterator here because of what you were telling me today of how io::Result has a built in Iterator and using that fact the code implicitly is walking the stack ?